src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/OleVariant.cs

```diff
@@ -0,0 +1,560 @@
+                variant._typeUnion._unionTypes._ui8 = (ulong)(object)value;
+            }
+            else
+            {
+                throw new ArgumentException("Unsupported type", nameof(T));
+            }
+            // We do not support mapping nint or nuint to VT_INT and VT_UINT respectively
+            // as this does not match the MS-OAUT spec.
+            // We do not map VT_BYREF automatically, nor do we map any of the array types.
+            return variant;
+        }
+#pragma warning restore CS0618
+
+        /// <summary>
+        /// Create a <see cref="OleVariant"/> with the given type and provided value.
+        /// </summary>
+        /// <typeparam name="T">The type of the value to store in the variant.</typeparam>
+        /// <param name="vt">The type of the variant</param>
+        /// <param name="rawValue">The raw value to store in the variant without any processing</param>
+        /// <returns>A variant that contains the provided value.</returns>
+        /// <exception cref="ArgumentException">When the provided <paramref name="vt"/> corresponds to a variant type that is not supported in VARIANTs or is <see cref="VarEnum.VT_DECIMAL"/></exception>
+        /// <exception cref="PlatformNotSupportedException">When the provided <paramref name="vt"/> specifies the <see cref="VarEnum.VT_ARRAY"/> flag for SAFEARRAYs.</exception>
+        public static unsafe OleVariant CreateRaw<T>(VarEnum vt, T rawValue)
+            where T : unmanaged
+        {
+            ArgumentOutOfRangeException.ThrowIfGreaterThan(Unsafe.SizeOf<T>(), sizeof(UnionTypes), nameof(T));
+            if (vt == VarEnum.VT_DECIMAL)
+            {
+                throw new ArgumentException("VT_DECIMAL is not supported in CreateRaw. Use the Create method.", nameof(vt));
+            }
+            if (vt == VarEnum.VT_VARIANT)
+            {
+                throw new ArgumentException("VT_VARIANT is not supported in variants.", nameof(vt));
+            }
+            if (vt.HasFlag(VarEnum.VT_ARRAY) && !OperatingSystem.IsWindows())
+            {
+                throw new PlatformNotSupportedException("OleVariants containing SAFEARRAYs are not supported on this platform");
+            }
+
+            Unsafe.SkipInit(out OleVariant value);
+            value.VarType = vt;
+            value.GetRawDataRef<T>() = (vt, sizeof(T)) switch
+            {
+                (VarEnum.VT_I1 or VarEnum.VT_UI1, 1) => rawValue,
+                (VarEnum.VT_I2 or VarEnum.VT_UI2 or VarEnum.VT_BOOL, 2) => rawValue,
+                (VarEnum.VT_ERROR or VarEnum.VT_HRESULT or VarEnum.VT_I4 or VarEnum.VT_UI4 or VarEnum.VT_R4 or VarEnum.VT_INT or VarEnum.VT_UINT, 4) => rawValue,
+                (VarEnum.VT_I8 or VarEnum.VT_UI8 or VarEnum.VT_R8 or VarEnum.VT_DATE, 8) => rawValue,
+                (VarEnum.VT_INT or VarEnum.VT_UINT or VarEnum.VT_UNKNOWN or VarEnum.VT_DISPATCH or VarEnum.VT_LPSTR or VarEnum.VT_BSTR or VarEnum.VT_LPWSTR or VarEnum.VT_SAFEARRAY
```

Comment:
`VarEnum.VT_INT or VarEnum.VT_UINT` are we trying to handle the `IntPtr` case here? If so, that warrants a comment and a specific test.

Reply:
This was a typo. I didn't mean to leave these options in this case. Looks like I'm missing some negative tests that I should add.